codalab/server/bundle_manager.py

```diff
@@ -320,9 +320,11 @@ def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run):
             reverse=True,
         )
 
-        # Get bundles in RUNNING state from the bundle table
-        running_bundles = self._model.batch_get_bundles(state=State.RUNNING)
```

Comment:
There is a race condition when fetching bundles in `state=running`. When a bundle moves from `staged` to `running` state, it took around 3 seconds to get it back from `batch_get_bundles` function. If `deduct_bundle_resources` happened during this 3 seconds, the resource calculation will be wrong. Switching to fetching `run_uuids` will be more accurate.

Reply:
I mean by the time we assign a bundle to run on a worker successfully in [Line 351](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L351), that bundle's `uuid` has been added to `workrer['run_uuid']`'s in-memory object. Then in the next round of iteration in [Line 324](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L324), the fetched results from the database don't have that bundle reflected in `running` state yet. However, since that bundle has already been added to `worker['run_uuid']` and we don't get its information from the bundle table, when we do resource deduction in [Line 362](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L362), it will just be skipped and the returned `workers_list` in [Line 372](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L372) will eventually end up having more resources left once the calculation is done. Then those workers will be assigned more jobs than their real capacity eventually.

Reply:
With the new SQL statement, it won't look for bundles in `running` state anymore. Instead, it will only search for bundles that exist in `run_uuid` and ignore the state condition check, which increases the chance to be able to fetch the actual bundle back. Then during the resource deduction, skipping checks for one bundle in [Line 362](https://github.com/codalab/codalab-worksheets/blob/master/codalab/server/bundle_manager.py#L362) won't happen frequently and hence increase the resource calculation accuracy.